static/src/javascripts/projects/common/modules/commercial/track-ad.js

```diff
@@ -0,0 +1,41 @@
+define([
+    'Promise',
+    'common/utils/mediator'
+], function (
+    Promise,
+    mediator
+) {
+    var promises;
+
+    function TrackAd() {
+        promises = {};
+    }
+
+    TrackAd.prototype.track = function (id) {
+        promises[id] = promises[id] || new Promise(function (resolve, reject) {
+            var onAdLoaded = function (event) {
+                if (event.slot.getSlotElementId() === id) {
+                    unlisten();
+                    resolve(!event.isEmpty);
+                }
+            };
+
+            var onAllAdsLoaded = function () {
+                unlisten();
+                reject(new Error('Slot ' + id + ' was never loaded'));
+            };
+
+            function unlisten() {
+                mediator.off('modules:commercial:dfp:rendered', onAdLoaded);
+                mediator.off('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+            }
+
+            mediator.on('modules:commercial:dfp:rendered', onAdLoaded);
+            mediator.on('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+        });
+
+        return promises[id];
+    };
+
+    return new TrackAd();
```

Comment:
wOOt. Can you show me an example?

On 2 March 2016 at 15:30, Oliver Joseph Ash <notifications@github.com>
wrote:

> In static/src/javascripts/projects/common/modules/commercial/track-ad.js
> <https://github.com/guardian/frontend/pull/12138#discussion_r54738480>:
>
> > +                reject(new Error('Slot ' + id + ' was never loaded'));
> > +            };
> > +
> > +            function unlisten() {
> > +                mediator.off('modules:commercial:dfp:rendered', onAdLoaded);
> > +                mediator.off('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
> > +            }
> > +
> > +            mediator.on('modules:commercial:dfp:rendered', onAdLoaded);
> > +            mediator.on('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
> > +        });
> > +
> > +        return promises[id];
> > +    };
> > +
> > +    return new TrackAd();
>
> Ah yes. In our tests we do use Squire which is supposed to help here,
> allowing you to import a module into multiple contexts, each with their own
> state.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/12138/files#r54738480>.
>



-- 
–Regis

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396